engine/access/rpc/connection/manager.go

```diff
@@ -0,0 +1,226 @@
+package connection
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/keepalive"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow-go/module"
+)
+
+// DefaultClientTimeout is used when making a GRPC request to a collection node or an execution node.
+const DefaultClientTimeout = 3 * time.Second
+
+// Manager provides methods for getting and managing gRPC client connections.
+type Manager struct {
+	cache      *Cache
+	logger     zerolog.Logger
+	metrics    module.AccessMetrics
+	maxMsgSize uint
+}
+
+// NewManager creates a new Manager with the specified parameters.
+func NewManager(
+	cache *Cache,
+	logger zerolog.Logger,
+	metrics module.AccessMetrics,
+	maxMsgSize uint,
+) Manager {
+	return Manager{
+		cache:      cache,
+		logger:     logger,
+		metrics:    metrics,
+		maxMsgSize: maxMsgSize,
+	}
+}
+
+// GetConnection returns a gRPC client connection for the given grpcAddress and timeout.
+// If a cache is used, it retrieves a cached connection, otherwise creates a new connection.
+// It returns the client connection and an io.Closer to close the connection when done.
+func (m *Manager) GetConnection(grpcAddress string, timeout time.Duration) (*grpc.ClientConn, io.Closer, error) {
+	if m.cache != nil {
+		conn, err := m.retrieveConnection(grpcAddress, timeout)
+		if err != nil {
+			return nil, nil, err
+		}
+		return conn, &noopCloser{}, err
+	}
+
+	conn, err := m.createConnection(grpcAddress, timeout, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return conn, io.Closer(conn), nil
+}
+
+// Remove removes the gRPC client connection associated with the given grpcAddress from the cache.
+// It returns true if the connection was removed successfully, false otherwise.
+func (m *Manager) Remove(grpcAddress string) bool {
```

Comment:
I think we need to obtain lock here to make sure that `res.ClientConn` is indeed initialized.
Without it we can run in scenario:
- GetOrAdd creates client, takes lock, but haven't created connection yet
- Another goroutine calls `Remove`, which gets a client in half-initialized state and removes it from cache.
- since `ClientConn` is not yet set, we will perform `Close` on nil `grpc.Connection`

Obtaining lock in remove should fix the problem

Reply:
good catch